src/pybind/mgr/drain_osd/module.py

```diff
@@ -0,0 +1,329 @@
+
+    def reweight_osds(self, osd_ids: Set[int]) -> bool:
+        results = [(self.reweight_osd(osd_id)) for osd_id in osd_ids]
+        return all(results)
+
+    def get_pg_count(self, osd_id: int, osd_df: Optional[dict] = None) -> int:
+        if not osd_df:
+            osd_df = self.osd_df()
+        osd_nodes = osd_df.get('nodes', [])
+        for osd_node in osd_nodes:
+            if osd_node.get('id', None) == int(osd_id):
+                return osd_node.get('pgs')
+        errmsg = f"Could not find <pgs> field for osd_id: {osd_id} in osd_df data"
+        self.log.error(errmsg)
+        raise RuntimeError(errmsg)
+
+    def get_osd_weight(self, osd_id: int) -> float:
+        osd_df = self.osd_df()
+        osd_nodes = osd_df.get('nodes', [])
+        for osd_node in osd_nodes:
+            if osd_node.get('id', None) == int(osd_id):
+                return float(osd_node.get('crush_weight'))
+        return -1.0
+
+    def reweight_osd(self, osd_id: int, weight: float = 0.0) -> bool:
+        if self.get_osd_weight(osd_id) == weight:
+            self.log.debug(f"OSD <{osd_id}> is already weighted with: {weight}")
+            return True
+        base_cmd = 'osd crush reweight'
+        ret, out, err = self.mon_command({
+            'prefix': base_cmd,
+            'name': f'osd.{osd_id}',
+            'weight': weight
+        })
+        cmd = f"{base_cmd} on osd.{osd_id} to weight: {weight}"
+        self.log.debug(f"running cmd: {cmd}")
+        if ret != 0:
+            self.log.error(f"command: {cmd} failed with: {err}")
+            return False
+        self.log.info(f"command: {cmd} succeeded with: {out}")
+        return True
+
+    def find_osd_stop_threshold(self, osd_ids: Set[int]) -> Set[int]:
+        """
+        Cut osd_id list in half until it's ok-to-stop
+
+        :param osd_ids: list of osd_ids
+        :return: list of ods_ids that can be stopped at once
+        """
+        if not osd_ids:
+            return set()
+        _osds: List[int] = list(osd_ids.copy())
+        while not self.ok_to_stop(_osds):
+            if len(_osds) <= 1:
+                # can't even stop one OSD, aborting
+                self.log.info("Can't even stop one OSD. Cluster is probably busy. Retrying later..")
+                return set()
+            self.event.wait(1)
+            # splitting osd_ids in half until ok_to_stop yields success
+            # maybe popping ids off one by one is better here..depends on the cluster size I guess..
+            # There's a lot of room for micro adjustments here
+            _osds = _osds[len(_osds)//2:]
+        return set(_osds)
+
+    def ok_to_stop(self, osd_ids: List[int]) -> bool:
+        base_cmd = "osd ok-to-stop"
+        ret, out, err = self.mon_command({
+            'prefix': base_cmd,
+            # apparently ok-to-stop allows strings only
+            'ids': [str(x) for x in osd_ids]
+        })
+        self.log.debug(f"running cmd: {base_cmd} on ids {osd_ids}")
```

Comment:
It's better to place this log before the `self.mon_command()`, and setting it to `info` level.